lms/djangoapps/certificates/api.py

```diff
@@ -484,49 +484,46 @@ def get_active_web_certificate(course, is_preview_mode=None):
     return None
 
 
-def get_certificate_template(course_key, mode, language):  # pylint: disable=unused-argument
+def get_certificate_template(course_key, mode, language):
     """
     Retrieves the custom certificate template based on course_key, mode, and language.
     """
-    org_id, template = None, None
     # fetch organization of the course
-    course_organization = get_course_organizations(course_key)
-    if course_organization:
-        org_id = course_organization[0]['id']
-
-    if org_id and mode:
-        template = CertificateTemplate.objects.filter(
-            organization_id=org_id,
-            course_key=course_key,
-            mode=mode,
-            is_active=True
-        )
-    # if don't template find by org and mode
-    if not template and org_id and mode:
-        template = CertificateTemplate.objects.filter(
-            organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
-        )
-    # if don't template find by only org
-    if not template and org_id:
-        template = CertificateTemplate.objects.filter(
-            organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=None,
-            is_active=True
-        )
-    # if we still don't template find by only course mode
-    if not template and mode:
-        template = CertificateTemplate.objects.filter(
-            organization_id=None,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
-        )
+    org_id = get_course_organization_id(course_key)
+
+    # prepare template sets.
+    active_templates = CertificateTemplate.objects.filter(is_active=True)
+    org_templates = active_templates.filter(organization_id=org_id)
+    mode_templates = active_templates.filter(mode=mode)
+    org_and_mode_templates = org_templates & mode_templates
+    template_by_key = org_and_mode_templates.filter(course_key=course_key)
+
+    # Look up templates from most specific to least specific.
+    # At each level check for a langauge specific template.
+    if template_by_key:  # check for template with correct course key, org, and mode
+        template_by_key = check_for_language_specific_template(template_by_key, language)
+        return template_by_key[0].template
+    elif org_and_mode_templates:  # check for template with the correct org and mode
+        org_and_mode_templates = check_for_language_specific_template(org_and_mode_templates, language)
+        return org_and_mode_templates[0].template
+    elif org_templates:  # check for template with only correct org
+        org_templates = check_for_language_specific_template(org_templates, language)
+        return org_templates[0].template
+    elif mode_templates:  # check for template with only correct mode
+        mode_templates = check_for_language_specific_template(mode_templates, language)
+        return mode_templates[0].template
```

Comment:
This is the common case and the result is that it takes 6 DB queries to get here. 4 in the conditional checks, and then 2 queries if there is no language template.  We could reduce the number of queries slightly by calling the `check_for_language_specific_template` method and dropping into the conditional if that returns something.